libs/ml/examples/centralised_distributed_learning/main.cpp

```diff
@@ -83,10 +83,9 @@ class TrainingClient
         g_.SetInput("Input", input.second.at(0));
         g_.SetInput("Label", input.first);
 
-        ArrayType loss_tensor = g_.Evaluate("Error").Copy();
-        loss += fetch::math::statistics::Mean(loss_tensor);
-        loss_tensor.Fill(static_cast<DataType>(1));
-        g_.BackPropagate("Error", loss_tensor);
+        ArrayType loss_tensor = g_.Evaluate("Error");
```

Comment:
it would be nice if we didn't have to do this. It's inefficient and ugly to generate a tensor as output from loss function when we know it should only be a DataType.

I'm not 100% sure the best way to achieve this, and maybe it should go in the next PR but how about something like this:
1. all ops that are loss functions have a flag/enum indicating that they are a loss type or not (we wanted to do this anyway to protect against using the output from the loss function in forward layers, or back-propping through the ground truth input)
2. Evaluate calls one of two different functions (EvalImplementation) depending on whether "Error" refers to a loss function op or not. Return type could be templated?

Not sure if that will work or not, but maybe we can think of something like this that makes sense